src/webview/__tests__/generateInboundEventEditSequence-test.js

```diff
@@ -0,0 +1,625 @@
+        // All together, sorted by ID. (Which basically means jumbled;
+        // the IDs in each sub-list are only internally sorted.)
+        messages: [...pmMessages1, ...pmMessages2, ...pmMessages3].sort((a, b) => a.id - b.id),
+      },
+    ].forEach(testCase => check(testCase));
+  });
+
+  const pmNarrowGroup = pmNarrowFromUsersUnsafe([user1, user2, user3]);
+  test(`${keyFromNarrow(pmNarrowGroup)}`, () => {
+    [
+      { narrow: pmNarrowGroup, messages: pmMessages4 },
+      { narrow: pmNarrowGroup, messages: pmMessages5 },
+      { narrow: pmNarrowGroup, messages: pmMessages6 },
+      {
+        narrow: pmNarrowGroup,
+        // All together, sorted by ID. (Which basically means jumbled;
+        // the IDs in each sub-list are only internally sorted.)
+        messages: [...pmMessages4, ...pmMessages5, ...pmMessages6].sort((a, b) => a.id - b.id),
+      },
+    ].forEach(testCase => check(testCase));
+  });
+
+  test(`${keyFromNarrow(ALL_PRIVATE_NARROW)}`, () => {
+    [
+      { narrow: ALL_PRIVATE_NARROW, messages: pmMessages1 },
+      { narrow: ALL_PRIVATE_NARROW, messages: pmMessages2 },
+      { narrow: ALL_PRIVATE_NARROW, messages: pmMessages3 },
+      { narrow: ALL_PRIVATE_NARROW, messages: pmMessages4 },
+      { narrow: ALL_PRIVATE_NARROW, messages: pmMessages5 },
+      { narrow: ALL_PRIVATE_NARROW, messages: pmMessages6 },
+      {
+        narrow: ALL_PRIVATE_NARROW,
+        // All together, sorted by ID. (Which basically means jumbled;
+        // the IDs in each sub-list are only internally sorted.)
+        messages: [
+          ...pmMessages1,
+          ...pmMessages2,
+          ...pmMessages3,
+          ...pmMessages4,
+          ...pmMessages5,
+          ...pmMessages6,
+        ].sort((a, b) => a.id - b.id),
+      },
+    ].forEach(testCase => check(testCase));
+  });
+});
+
+describe('getEditSequence correct for interesting changes', () => {
+  const resetMsglist = () => {
+    invariant(document.body, 'expected jsdom environment');
+    document.body.innerHTML = '<div id="msglist-elements" />';
+  };
+
+  const getClonedMsglistElementsDiv = () => {
+    const msglistElementsDiv = document.querySelector('div#msglist-elements');
+    invariant(msglistElementsDiv, 'getClonedMsglistElementsDiv: expected msglistElementsDiv');
+
+    return msglistElementsDiv.cloneNode(true);
+  };
+
+  const check = (
+    { oldBackgroundData = eg.backgroundData, oldNarrow = HOME_NARROW, oldMessages },
+    { newBackgroundData = eg.backgroundData, newNarrow = HOME_NARROW, newMessages },
+    expectedEditSequenceLength?: number,
+  ) => {
+    const oldElements = getMessageListElements(oldMessages, oldNarrow);
+    const newElements = getMessageListElements(newMessages, newNarrow);
+
+    resetMsglist();
+
+    applyEditSequence(
+      getEditSequence(
+        {
+          backgroundData: newBackgroundData,
+          narrow: newNarrow,
+          elements: [],
+          _: m => m,
+        },
```

Comment:
nit: This object can be collapsed to one line. (Quick way to edit: delete the newline after the `{`, then autoformat.)

That makes this section of code 30 lines shorter, making it significantly easier to look at its logic all at once. I'd probably then also delete some of the blank lines to group things together a bit more, like between each `resetMsgList` and the following `applyEditSequence`.